File: 0020-valid-parentheses/solution.py
class Solution:
    def __init__(self):
        self.stack = []
    
    def pop(self):
        return self.stack.pop()
    
    def top(self):
        return self.stack[-1]
    
    def push(self, value):
        self.stack.append(value)
        
    def isValid(self, s: str) -> bool:
        result = False
        
        for bracket in s:
            if bracket in ["(", "{", "["]:
                self.push(bracket)
            else:
                if len(self.stack) == 0:
                    return False
                t = self.pop()
                if (t == "{" and bracket != "}") or ( t == "[" and bracket != "]") or (t == "(" and bracket != ")"):
                    return False
        if len(self.stack) != 0:
            return False
        else:
            return True
        

File: 0020-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> parenStack;

        for (int i = 0; i < s.size(); ++i) {
            if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
                parenStack.push(s[i]);
            } else {
                char closeParen = '(';
                if (s[i] == '}') closeParen = '{';
                else if (s[i] == ']') closeParen = '[';

                if (parenStack.size() && parenStack.top() == closeParen)
                    parenStack.pop();
                else
                    return false;
            }
        }

        return parenStack.size() == 0;
    }
};
